application/src/main/java/org/togetherjava/tjbot/commands/moderation/modmail/ModMailCommand.java

```diff
@@ -0,0 +1,178 @@
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Implements the '/modmail' command which can be used to automatically redirects messages from DM
+ * or Guild.
+ * <p>
+ * Actions are then logged to a dedicated channel, given by
+ * {@link Config#getModAuditLogChannelPattern()}.
+ */
+
+public final class ModMailCommand extends SlashCommandAdapter {
+
+    private static final String COMMAND_NAME = "modmail";
+    private static final String MESSAGE = "message";
+    private static final String OPTION_MESSAGE_PRIVATE = "message-private";
+    private static final int COOLDOWN_DURATION_VALUE = 30;
+    private static final ChronoUnit COOLDOWN_DURATION_UNIT = ChronoUnit.MINUTES;
+    private final Cache<Long, Instant> channelIdToLastCommandInvocation;
+    private final Predicate<String> modMailChannelNamePredicate;
+    private final Predicate<String> guildNamePatternPredicate;
+
+
+    /**
+     * Creates a new instance.
+     *
+     * @param config the config to use for this
+     */
+    public ModMailCommand(Config config) {
+        super(COMMAND_NAME, "Sends a message to the moderators", SlashCommandVisibility.GLOBAL);
+        getData().addOption(OptionType.STRING, MESSAGE, "Message to the moderators", true)
+            .addOption(OptionType.BOOLEAN, OPTION_MESSAGE_PRIVATE,
+                    "do you wish for your message to stay private", false);
+
+        channelIdToLastCommandInvocation = createCooldownCache();
+
+        modMailChannelNamePredicate =
+                Pattern.compile(config.getModMailChannelPattern()).asMatchPredicate();
+        guildNamePatternPredicate =
+                Pattern.compile(config.getGuildNamePattern()).asMatchPredicate();
+    }
+
+    private Cache<Long, Instant> createCooldownCache() {
+        return Caffeine.newBuilder()
+            .maximumSize(1_000)
+            .expireAfterAccess(COOLDOWN_DURATION_VALUE, TimeUnit.of(COOLDOWN_DURATION_UNIT))
+            .build();
+    }
+
+    @Override
+    public void onSlashCommand(SlashCommandInteractionEvent event) {
+        MessageChannel messageChannel = event.getChannel();
+        if (isChannelOnCooldown(messageChannel)) {
+            event
+                .reply("Please wait a bit, this command can only be used once per %d %s.".formatted(
+                        getRemainingTime(messageChannel),
+                        COOLDOWN_DURATION_UNIT.toString().toLowerCase(Locale.US)))
+                .setEphemeral(true)
+                .queue();
+            return;
+        }
+        channelIdToLastCommandInvocation.put(messageChannel.getIdLong(), Instant.now());
+
+        Optional<TextChannel> auditLogChannel = getChannel(event);
+        if (auditLogChannel.isEmpty()) {
+            return;
+        }
```

Comment:
this needs some sort of log warn message. otherwise we dont notice that we messed up

Reply:
Done